pkg/columns/columninfo.go

```diff
@@ -0,0 +1,196 @@
+	fieldIndex      int          // used for the main struct
+	subFieldIndex   []int        // used for embedded structs
+	kind            reflect.Kind // cached kind info from reflection
+	columnType      reflect.Type // cached type info from reflection
+}
+
+func (ci *Column[T]) fromTag(tag string) error {
+	tagInfo := strings.Split(tag, ",")
+	ci.Name = tagInfo[0]
+
+	tagInfo = tagInfo[1:]
+	for _, subTag := range tagInfo {
+		params := strings.SplitN(subTag, ":", 2)
+		paramsLen := len(params)
+		switch params[0] {
+		case "align":
+			if paramsLen == 1 {
+				return fmt.Errorf("missing alignment value for field %q", ci.Name)
+			}
+			switch params[1] {
+			case "left":
+				ci.Alignment = AlignLeft
+			case "right":
+				ci.Alignment = AlignRight
+			default:
+				return fmt.Errorf("invalid alignment %q for field %q", params[1], ci.Name)
+			}
+		case "precision":
+			if ci.kind != reflect.Float32 && ci.kind != reflect.Float64 {
+				return fmt.Errorf("field %q is not a float field and thereby cannot have precision defined", ci.Name)
+			}
+			if paramsLen == 1 {
+				return fmt.Errorf("missing precision value for field %q", ci.Name)
+			}
+			w, err := strconv.Atoi(params[1])
+			if err != nil {
+				return fmt.Errorf("invalid precision value %q for field %q: %w", params[1], ci.Name, err)
+			}
+			if w < 0 {
+				return fmt.Errorf("negative precision value %q for field %q", params[1], ci.Name)
+			}
+			ci.Precision = w
+		case "ellipsis":
+			if paramsLen == 1 {
+				ci.EllipsisType = ellipsis.End
+				continue
+			}
+			switch params[1] {
+			case "end", "":
+				ci.EllipsisType = ellipsis.End
+			case "middle":
+				ci.EllipsisType = ellipsis.Middle
+			case "none":
+				ci.EllipsisType = ellipsis.None
+			case "start":
+				ci.EllipsisType = ellipsis.Start
+			default:
+				return fmt.Errorf("invalid ellipsis value %q for field %q", params[1], ci.Name)
+			}
+		case "fixed":
+			if paramsLen != 1 {
+				return fmt.Errorf("parameter fixed on field %q must not have a value", ci.Name)
+			}
+			ci.FixedWidth = true
+		case "group":
+			if paramsLen == 1 {
+				return fmt.Errorf("missing group value for field %q", ci.Name)
+			}
+			switch params[1] {
+			case "sum":
+				if !ci.columnType.ConvertibleTo(reflect.TypeOf(int(0))) {
+					return fmt.Errorf("cannot use sum on field %q of kind %q", ci.Name, ci.kind.String())
+				}
+				ci.GroupType = GroupTypeSum
+			default:
+				return fmt.Errorf("invalid sum value %q for field %q", params[1], ci.Name)
+			}
+		case "hide":
+			if paramsLen != 1 {
+				return fmt.Errorf("parameter hide on field %q must not have a value", ci.Name)
+			}
+			ci.Visible = false
+		case "width":
+			if paramsLen == 1 {
+				return fmt.Errorf("missing width value for field %q", ci.Name)
+			}
+			w, err := strconv.Atoi(params[1])
+			if err != nil {
+				return fmt.Errorf("invalid width %q for field %q: %w", params[1], ci.Name, err)
+			}
+			ci.Width = w
+		default:
```

Comment:
Great idea to handle it that way. I would probably change or at least default to WithoutKeywords() (maybe name it WithoutTags()), so everything is used by default. This should disable the given columns completely, then, so that even calls to GetColumns() wouldn't return them and you can't use sort etc on it. Maybe we should consider adding columnTags:"abc,def" instead of putting it inside the column:"..." to allow multiple keywords.

Btw: the order defaults to the order of fields in the struct (*10), but an optional "order:number" can be used to reorder it (haven't pushed that change, yet). And then you can always use SetShowColumns() on the formatter to force columns and the order.

Reply:
> I would probably change or at least default to WithoutKeywords() (maybe name it WithoutTags()), so everything is used by default.
>
I would prefer that the default behaviour is to return the gadget-specific columns. It means, those without keywords and without `hide`. IMO, It would be more intuitive for users.

> This should disable the given columns completely, then, so that even calls to GetColumns() wouldn't return them and you can't use sort etc on it. 
>
I am not sure I follow you. Could you please describe this idea in more detail?

> Maybe we should consider adding columnTags:"abc,def" instead of putting it inside the column:"..." to allow multiple keywords.
>
Yeah, it would be more readable than `column:"name,keyword:kubernetes,keyword:runtime,width:16"`.

---

Regarding the 4th use-case (`local-gadget --k8s-metadata`):

> Btw: the order defaults to the order of fields in the struct (*10), but an optional "order:number" can be used to reorder it (haven't pushed that change, yet).
>
But such order is defined in the struct, so there is no way to have a different order between two callers (`kubectl-gadget` and `local-gadget`).

> And then you can always use SetShowColumns() on the formatter to force columns and the order.
> 
Yeah, I was thinking about something like this:

```golang
func NewTraceParserWithRuntimeInfo(commonFlags *utils.CommonFlags) (TraceParser[types.Event], error) {
	keywords := []string{columns.KeywordContainerRuntime}
	if commonFlags.K8sMetadata {
		keywords = append(keywords, columns.KeywordKubernetes)
	}

	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(keywords...))
	if err != nil {
		return nil, fmt.Errorf("creating columns: %w", err)
	}

	var defColumns []string
	if commonFlags.K8sMetadata {
		// Based on keywords, define the defColumns:
		//  - First the columns with "runtime" keyword: Container runtime, ID, Container
		//  - Then, the columns without keyword (gadget-specific): PID, PPID, PCOMM...
		//  - Finally, the columns with "kubernetes" keyword: Pod, Namespace, Container. BTW, for these, we need to change the name and add the "K8S-" prefix.
		// Example:
		// $ local-gadget trace exec --k8s-metadata
		// CONTAINER RUNTIME    ID        CONTAINER       PID     PPID    PCOMM ... K8S-POD    K8S-NAMESPACE    K8S-CONTAINER
	}

	return textcolumns.NewFormatter(cols, &textcolumns.Options{DefaultColumns: defColumns}), nil
}
```

What do you think?

cc @mauriciovasquezbernal

Reply:
I noticed something about this proposal. If I am making this call:

```golang
	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(columns.KeywordKubernetes))
```

I would expect to get only the columns with `KeywordKubernetes` keyword but, I was supposing it will return also the columns without tags. Maybe that's why you meant with `WithoutKeyword`. By default, if no options are provided, it will return the column without keywords. Instead, if the user wants columns with Kubernetes and no keyword, they has to do:

```golang
	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(columns.KeywordKubernetes), columns.WithoutKeywords())
```

Reply:
I think we got this covered now.